lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,77 +3640,989 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+      // created must be the differential function's entry. Create differential
+      // entry arguments and continue.
+      if (&origBB == origEntry) {
+        assert(diffBB->isEntry());
+        createEntryArguments(&differential);
+        auto *lastArg = diffBB->getArguments().back();
+        assert(lastArg->getType() == diffStructLoweredType);
+        differentialStructArguments[&origBB] = lastArg;
+      }
+    }
+    assert(diffBBMap.size() == 1 &&
+           "Can only currently handle single basic block functions");
+
+    // The differential function has type:
+    // (arg0', ..., argn', exit_diffs) -> result'.
+    auto &diffBuilder = getDifferentialBuilder();
+    auto diffParamArgs =
+        differential.getArgumentsWithoutIndirectResults().drop_back();
+    assert(diffParamArgs.size() == attr->getIndices().parameters->getCapacity());
+    auto origParamArgs = original->getArgumentsWithoutIndirectResults();
+
+    // Check if result is not varied.
+    SmallVector<SILValue, 8> origFormalResults;
+    collectAllFormalResultsInTypeOrder(*original, origFormalResults);
+    auto origResult = origFormalResults[getIndices().source];
+    // Emit warning if original result is not varied, because it will always
+    // have a zero derivative.
+    if (!activityInfo.isVaried(origResult, getIndices().parameters)) {
+      // Emit fixit if original result has a valid source location.
+      auto startLoc = origResult.getLoc().getStartSourceLoc();
+      auto endLoc = origResult.getLoc().getEndSourceLoc();
+      if (startLoc.isValid() && endLoc.isValid()) {
+        context
+            .diagnose(startLoc, diag::autodiff_nonvaried_result_fixit)
+            .fixItInsert(startLoc, "withoutDerivative(at:")
+            .fixItInsertAfter(endLoc, ")");
+      }
+    }
+
+    auto *diffEntry = getDifferential().getEntryBlock();
+    auto diffLoc = getDifferential().getLocation();
+    diffBuilder.setInsertionPoint(
+        diffEntry, getNextDifferentialLocalAllocationInsertionPoint());
+
+    for (auto index : *getIndices().parameters) {
+      auto diffParam = diffParamArgs[index];
+      auto origParam = origParamArgs[index];
+
+      diffBuilder.createRetainValue(diffLoc, diffParam,
+                                    diffBuilder.getDefaultAtomicity());
+      setTangentValue(
+          origEntry, origParam, makeConcreteTangentValue(diffParam));
+
+      LLVM_DEBUG(getADDebugStream()
+                 << "Assigned parameter " << *diffParam
+                 << " as the tangent of original result " << origParam);
+    }
+
+    // Clone.
+    SmallVector<SILValue, 4> entryArgs(entry->getArguments().begin(),
+                                       entry->getArguments().end());
+    cloneFunctionBody(original, entry, entryArgs);
+    // If errors occurred, back out.
+    if (errorOccurred)
+      return true;
+
+    LLVM_DEBUG(getADDebugStream() << "Generated JVP for "
```

Comment:
Are generated differentials debug printed anywhere?
If not, could you please debug print it *after* printing the JVP?
```diff
+LLVM_DEBUG(getADDebugStream() << "Generated differential for "
+               << original->getName() << ":\n" << getDifferential());
```

Note that pullbacks are currently printed *before* the VJP, which I've found to be slightly suboptimal for debugging personally. No changes there necessary for now.

Reply:
I was confused at first as I swear as I was, except that was occuring in the other two PRs I'm working on. Will add it here!